### PR TITLE
Implement hover menu support for ruff-server; Issue #10595

### PR DESCRIPTION
**Pull request by nolanking90**
## Summary

Add support for hover menu to ruff_server, as requested in [10595](https://github.com/astral-sh/ruff/issues/10595).
Majority of new code is in hover.rs.
I reused the regex from ruff-lsp's implementation. Also reused the format_rule_text function from ruff/src/commands/rule.rs
Added capability registration in server.rs, and added the handler to api.rs.

## Test Plan

Tested in NVIM v0.10.0-dev-2582+g2a8cef6bd, configured with lspconfig using the default options (other than cmd pointing to my test build, with options "server" and "--preview"). OS: Ubuntu 24.04, kernel 6.8.0-22.

### CODE REVIEW

**Review by snowsignal — CHANGES_REQUESTED** [review]
Thank you for getting this done! You have the honor of being the first outside contributor for `ruff server` 😄 

There are a few things I'd like to improve before we merge this:

1. **Error resiliency** - there's a lot of `.unwrap()`s in here that could potentially panic in a real-world scenario. When I was testing this PR, I noticed that hovering over an incomplete `noqa` comment would cause the server to panic (for example: `# noqa: RUF`). I've made some suggestions that use early returns in the place of potentially panicking functions like `.unwrap()`.
1. **Code complexity** - sections of this implementation are actually more complicated than they need to be. I've left suggestions on how we can simplify them.

Once again, I appreciate you taking the time to implement this! This is going to be a nifty feature.

File: Cargo.lock [hunk 1]
```diff
@@ -347,7 +347,7 @@ dependencies = [
  "heck 0.5.0",
  "proc-macro2",
  "quote",
- "syn 2.0.60",
+ "syn 2.0.59",
```
[the remark this answers is not in the document]
Comment (nolanking90) [reply Cargo.lock]: I reverted the lock file and pushed a commit, but now it's failing the check about the cargo.lock file. I'm not sure what's going on here (I assume I'm doing something wrong in git). Will return to this after I finish the other fixes.

**Review by snowsignal — APPROVED** [review]
Excellent! This seems to be working great on VS Code. Good catch with the comparison operators! I have one small nit, but otherwise you are free to merge 👍

**Comment by snowsignal** [comment]
My bad, I actually have to run the merge. Thanks for bearing with me 🤦‍♀️

**Pull request merged**